ouster-ros/include/ouster_ros/sensor_diagnostics_tracker.h

```
@@ -106,4 +112,209 @@ class SensorDiagnosticsTracker final
+
+  diagnostic_msgs::msg::DiagnosticStatus get_current_status() const;
+
+  std::map<std::string, std::string> get_debug_context(
+    const std::string & sensor_hostname, bool sensor_connection_active) const;
+
+  diagnostic_msgs::msg::DiagnosticStatus create_diagnostic_status(
+    const std::string & message, diagnostic_msgs::msg::DiagnosticStatus::_level_type level,
+    const std::map<std::string, std::string> & debug_context = {}) const;
+
+  template <typename MsgT>
+  void record_msg(const std::string & topic_name, const MsgT & msgs);
+
+private:
+  void produce_diagnostics(diagnostic_updater::DiagnosticStatusWrapper & stat);
+
+  double get_packet_rate(uint32_t packet_count, const rclcpp::Time & start_time) const;
+  bool is_sensor_healthy() const;
+
+  rclcpp::Logger logger_;
+  std::unique_ptr<diagnostic_updater::Updater> updater_;
+
+  // Current sensor status
+  std::string sensor_hostname_{"unknown"};
+  bool sensor_connection_active_{false};
+
+  SensorDiagnosticsState base_;
+  DiagnosticsVisitorRegistryType msg_analyzer_;
+};
+
+// Template implementation for ROS nodes
+template <typename DiagnosticsVisitorRegistryType>
+template <typename NodeT>
+SensorDiagnosticsTracker<DiagnosticsVisitorRegistryType>::SensorDiagnosticsTracker(
+  const std::string & name, NodeT * node, const std::string & hardware_id,
+  const DiagnosticsVisitorRegistryType & msg_analyzer)
+: logger_(node->get_logger()),
+  updater_(nullptr),
+  base_(name, node->get_clock(), hardware_id),
+  msg_analyzer_(msg_analyzer)
+{
+  try {
+    updater_ = std::make_unique<diagnostic_updater::Updater>(node);
+    updater_->setHardwareID(base_.get_hardware_id());
+    updater_->add(name + " Status", this, &SensorDiagnosticsTracker::produce_diagnostics);
+    RCLCPP_INFO(logger_, "Diagnostic updater initialized for %s", name.c_str());
+  } catch (const std::exception & e) {
+    RCLCPP_WARN(logger_, "Failed to initialize diagnostic updater: %s", e.what());
```

Shouldnt be ERR?
Do we want to continue execution if constructor fails? Maybe rethrow the exception?